src/main/java/org/springframework/social/partnercenter/security/AzureADMultiTenantOAuthOperations.java

```diff
@@ -0,0 +1,41 @@
+package org.springframework.social.partnercenter.security;
+
+import org.springframework.social.oauth2.AccessGrant;
+import org.springframework.social.oauth2.OAuth2Parameters;
+import org.springframework.social.partnercenter.http.logging.LogLevel;
+import org.springframework.util.MultiValueMap;
+
+public interface AzureADMultiTenantOAuthOperations {
+    String buildAuthorizeUrl(String clientId, String redirectUri, String state);
+
+    String buildAuthorizeUrl(String clientId, OAuth2Parameters additionalParams);
+
+    /**
+     * Exchange the authorization code for a refresh token
+     * @param authorizationCode the authorization code returned by the provider upon user authorization
+     * @param partnerTenantId the tenantId of the tenant that is being authorized
+     * @param redirectUri the authorization callback url; this value must match the redirectUri registered with the provider
+     * @param additionalParameters any additional parameters to be sent when exchanging the authorization code for an access grant. Should not be encoded.
+     * @return the access grant.
+     */
+    String exchangeForRefresh(String authorizationCode, String partnerTenantId, String redirectUri, MultiValueMap<String, String> additionalParameters);
```

Comment:
exchangeForRefreshToken maybe?

Reply:
or just exchangeForAccess and return the AccessGrant. The access grant returned I presume would work on the partner's tenant.

Reply:
According to Krishna, only the refresh token should be usable. The accessToken wouldn't work on partner center or graph because it wasn't obtained against a specific resource. That's why I am returning only the refreshToken.